Mathlib/NumberTheory/EllipticDivisibilitySequence.lean

```diff
@@ -189,43 +272,76 @@ lemma normEDS'_even (m : ℕ) : normEDS' b c d (2 * (m + 3)) * b =
     ring1
 
 /-- The canonical example of a normalised EDS `W : ℤ → R`,
-with initial values `W(0) = 0`, `W(1) = 1`, `W(2) = b`, `W(3) = c`, and `W(4) = b * d`.
+with initial values `W(0) = 0`, `W(1) = 1`, `W(2) = b`, `W(3) = c`, and `W(4) = d * b`.
 
 This extends `normEDS'` by defining its values at negative integers. -/
-def normEDS (n : ℤ) : R := n.sign * normEDS' b c d n.natAbs
+def normEDS (n : ℤ) : R :=
+  n.sign * normEDS' b c d n.natAbs
+
+@[simp]
+lemma normEDS_ofNat (n : ℕ) : normEDS b c d n = normEDS' b c d n := by
+  by_cases hn : n = 0
+  · rw [hn, normEDS, Nat.cast_zero, Int.sign_zero, Int.cast_zero, zero_mul, normEDS'_zero]
+  · rw [normEDS, Int.sign_coe_nat_of_nonzero hn, Int.cast_one, one_mul, Int.natAbs_cast]
 
 @[simp]
 lemma normEDS_zero : normEDS b c d 0 = 0 := by
-  erw [normEDS, Int.cast_zero, zero_mul]
+  erw [normEDS_ofNat, normEDS'_zero]
 
 @[simp]
 lemma normEDS_one : normEDS b c d 1 = 1 := by
-  erw [normEDS, Int.cast_one, one_mul, normEDS'_one]
+  erw [normEDS_ofNat, normEDS'_one]
 
 @[simp]
 lemma normEDS_two : normEDS b c d 2 = b := by
-  erw [normEDS, Int.cast_one, one_mul, normEDS'_two]
+  erw [normEDS_ofNat, normEDS'_two]
 
 @[simp]
 lemma normEDS_three : normEDS b c d 3 = c := by
-  erw [normEDS, Int.cast_one, one_mul, normEDS'_three]
+  erw [normEDS_ofNat, normEDS'_three]
 
 @[simp]
 lemma normEDS_four : normEDS b c d 4 = d * b := by
-  erw [normEDS, Int.cast_one, one_mul, normEDS'_four]
+  erw [normEDS_ofNat, normEDS'_four]
 
 lemma normEDS_odd (m : ℕ) : normEDS b c d (2 * (m + 2) + 1) =
     normEDS b c d (m + 4) * normEDS b c d (m + 2) ^ 3 -
       normEDS b c d (m + 1) * normEDS b c d (m + 3) ^ 3 := by
-  repeat erw [normEDS, Int.cast_one, one_mul]
-  exact normEDS'_odd b c d m
+  repeat erw [normEDS_ofNat]
+  exact normEDS'_odd ..
 
 lemma normEDS_even (m : ℕ) : normEDS b c d (2 * (m + 3)) * b =
     normEDS b c d (m + 2) ^ 2 * normEDS b c d (m + 3) * normEDS b c d (m + 5) -
       normEDS b c d (m + 1) * normEDS b c d (m + 3) * normEDS b c d (m + 4) ^ 2 := by
-  repeat erw [normEDS, Int.cast_one, one_mul]
-  exact normEDS'_even b c d m
+  repeat erw [normEDS_ofNat]
+  exact normEDS'_even ..
 
 @[simp]
 lemma normEDS_neg (n : ℕ) : normEDS b c d (-n) = -normEDS b c d n := by
-  simp [normEDS]
+  rw [normEDS, Int.sign_neg, Int.cast_neg, neg_mul, Int.natAbs_neg, normEDS]
+
+/-- Strong recursion principle for a normalised EDS indexed by `ℕ`: if we have
+ * `P 0`, `P 1`, `P 2`, `P 3`, and `P 4`,
+ * for all `m : ℕ` we can prove `P (2 * (m + 3))` from `P k` for all `k < 2 * (m + 3)`, and
+ * for all `m : ℕ` we can prove `P (2 * (m + 2) + 1)` from `P k` for all `k < 2 * (m + 2) + 1`,
+then we have `P n` for all `n : ℕ`. -/
+@[elab_as_elim]
+noncomputable def normEDSRec' {P : ℕ → Sort u} (base0 : P 0) (base1 : P 1) (base2 : P 2)
+    (base3 : P 3) (base4 : P 4) (even : ∀ m : ℕ, (∀ k < 2 * (m + 3), P k) → P (2 * (m + 3)))
+    (odd : ∀ m : ℕ, (∀ k < 2 * (m + 2) + 1, P k) → P (2 * (m + 2) + 1)) (n : ℕ) : P n :=
```

Comment:
I wonder what's your consideration for including all smaller `k` rather than just what's necessary, for example
```suggestion
    (odd : ∀ m : ℕ, P (m + 1) → P (m + 2) → P (m + 3) → P (m + 4) → P (2 * (m + 2) + 1)) (n : ℕ) : P n :=
```
in this case? Seems like this would save you some `linarith` when applying it.

Reply:
Another comment: since this is Sort-valued, you could even define `preNormEDS'` using this, or directly define `preNormEDS` using `normEDSRec` below.

Reply:
I will change the assumptions. I was told that defining `preNormEDS` using the equation compiler is more elegant.

Reply:
Yeah I already changed the `even` `odd` assumptions as indicated above in #13057. I agree it's probably fine to leave the definition of `preNormEDS'` as is.

Reply:
Actually after thinking about it for a bit maybe including all smaller `k` might make it more convenient for me in #10878? In line 464 of the `DivisionPolynomial` file I did something fancy: I replaced the inductive hypotheses with something in terms of `natDegree` and `coeff`, but the same chunk of code somehow worked in the even and odd case simultaneously! If I used your version I feel like the code would be twice longer.

Reply:
We can certainly have both versions if that's the case :)